File: operators.c
#include <stdio.h>

int main()
{
    int a = 5;
    int b = 4;

    // Arithmetic Operators
    printf("Addition: %d\n", a + b);
    printf("Subtraction: %d\n", a - b);
    printf("Multiplication: %d\n", a * b);
    printf("Division: %d\n", a / b);
    printf("Modulo: %d\n\n", a % b);

    // Comparison/Relational Operators
    printf("Equal to: %d\n", a == b);
    printf("Not equal to: %d\n", a != b);
    printf("Greater than: %d\n", a > b);
    printf("Less than: %d\n", a < b);
    printf("Greater than or equal to: %d\n", a >= b);
    printf("Less than or equal to: %d\n\n", a <= b);

    // Logical Operators
    printf("Logical NOT: %d\n", !(a > b));
    printf("Logical AND: %d\n", (a > b) && (a < b));
    printf("Logical OR: %d\n\n", (a > b) || (a < b));

    // Bitwise Operators
    printf("Bitwise NOT: %d\n", ~a);
    printf("Bitwise AND: %d\n", a & b);
    printf("Bitwise OR: %d\n", a | b);
    printf("Bitwise XOR: %d\n", a ^ b);
    printf("Bitwise Left Shift: %d\n", a << 2);
    printf("Bitwise Right Shift: %d\n\n", a >> 2);

    // Augmented/Compound Arithmetic Assignment Operators
    a += 2;
    printf("Addition Assignment: %d\n", a);

    a -= 4;
    printf("Subtraction Assignment: %d\n", a);

    a *= 5;
    printf("Multiplication Assignment: %d\n", a);

    a /= 4;
    printf("Division Assignment: %d\n", a);

    a %= 2;
    printf("Modulo Assignment: %d\n\n", a);

    // Augmented/Compound Bitwise Assignment Operators
    b &= 7;
    printf("Bitwise AND Assignment: %d\n", b);

    b |= 8;
    printf("Bitwise OR Assignment: %d\n", b);

    b ^= 10;
    printf("Bitwise XOR Assignment: %d\n", b);

    b <<= 2;
    printf("Bitwise Left Shift Assignment: %d\n", b);

    b >>= 3;
    printf("Bitwise Right Shift Assignment: %d", b);

    return 0;
}

// Truth Table {ref: 2}
// +---+---+------+-----+------+
// | p | q | AND  | OR  | XOR  |
// +---+---+------+-----+------+
// | 1 | 1 | 1    | 1   | 0    |
// | 1 | 0 | 0    | 1   | 1    |
// | 0 | 1 | 0    | 1   | 1    |
// | 0 | 0 | 0    | 0   | 0    |
// +---+---+------+-----+------+

// Rules
// 1. Do not use the ternary conditional operator, as it decreases readability
//    and is language-dependent.

// References
// 1. https://en.wikipedia.org/wiki/Operators_in_C_and_C++
// 2. https://en.wikipedia.org/wiki/Truth_table

File: operators.py
def main() -> None:
    a: int = 5
    b: int = 4

    # Arithmetic Operators
    print(f"Addition: {a + b:d}")
    print(f"Subtraction: {a - b:d}")
    print(f"Multiplication: {a * b:d}")
    print(f"Division: {a // b:d}")
    print(f"Modulo: {a % b:d}\n")

    # Comparison/Relational Operators
    print(f"Equal to: {a == b:d}")
    print(f"Not equal to: {a != b:d}")
    print(f"Greater than: {a > b:d}")
    print(f"Less than: {a < b:d}")
    print(f"Greater than or equal to: {a >= b:d}")
    print(f"Less than or equal to: {a <= b:d}\n")

    # Logical Operators
    print(f"Logical NOT: {not (a > b):d}")
    print(f"Logical AND: {(a > b) and (a < b):d}")
    print(f"Logical OR: {(a > b) or (a < b):d}\n")

    # Bitwise Operators
    print(f"Bitwise NOT: {~a:d}")
    print(f"Bitwise AND: {a & b:d}")
    print(f"Bitwise OR: {a | b:d}")
    print(f"Bitwise XOR: {a ^ b:d}")
    print(f"Bitwise Left Shift: {a << 2:d}")
    print(f"Bitwise Right Shift: {a >> 2:d}\n")

    # Augmented/Compound Arithmetic Assignment Operators
    a += 2
    print(f"Addition Assignment: {a:d}")

    a -= 4
    print(f"Subtraction Assignment: {a:d}")

    a *= 5
    print(f"Multiplication Assignment: {a:d}")

    a //= 4
    print(f"Division Assignment: {a:d}")

    a %= 2
    print(f"Modulo Assignment: {a:d}\n")

    # Augmented/Compound Bitwise Assignment Operators
    b &= 7
    print(f"Bitwise AND Assignment: {b:d}")

    b |= 8
    print(f"Bitwise OR Assignment: {b:d}")

    b ^= 10
    print(f"Bitwise XOR Assignment: {b:d}")

    b <<= 2
    print(f"Bitwise Left Shift Assignment: {b:d}")

    b >>= 3
    print(f"Bitwise Right Shift Assignment: {b:d}")


if __name__ == "__main__":
    main()

# Truth Table {ref: 1}
# +---+---+------+-----+------+
# | p | q | AND  | OR  | XOR  |
# +---+---+------+-----+------+
# | T | T | T    | T   | F    |
# | T | F | F    | T   | T    |
# | F | T | F    | T   | T    |
# | F | F | F    | F   | F    |
# +---+---+------+-----+------+

# Rules
# 1. Do not use the ternary conditional operator, as it decreases readability
#    and is language-dependent.
# 2. Do not use the exponentiation operator, as it is language-dependent.
# 3. Use the floor division when both the operands are of type int;
#    otherwise, use the regular division if any operand is of type float.
# 3.1. The regular division produces an accurate result of type float,
#      regardless of any given type of operands.
# 3.2. The floor division rounds the result down to the nearest whole number,
#      which can be of type int or float when any given operand is of type
#      float.

# References
# 1. https://en.wikipedia.org/wiki/Truth_table
